cachito/web/content_manifest.py

```diff
@@ -21,21 +21,54 @@ def __init__(self, request=None):
         self.request = request
         # dict to store go package level data; uses the package purl as key to identify a package
         self._gopkg_data = {}
+        # dict to store go module level purl dependencies. Module names are used as keys
+        self._gomod_data = {}
+
+    def process_gomod(self, package, dependency):
+        """
+        Process gomod package.
+
+        :param Package package: the gomod package to process
+        :param Dependency dependency: the gomod package dependency to process
+        """
+        if dependency.type == "gomod":
+            icm_source = {"purl": dependency.to_purl()}
+            self._gomod_data[package.name].append(icm_source)
 
     def process_go_package(self, package, dependency):
         """
-        Process gomod or go-package package.
+        Process go-package package.
 
-        :param Package package: the gomod or go-package package to process
-        :param Dependency dependency: the gomod or go-package package dependency to process
+        :param Package package: the go-package package to process
+        :param Dependency dependency: the go-package package dependency to process
         """
         purl = package.to_purl()
         if dependency.type == "go-package":
             icm_dependency = {"purl": dependency.to_purl()}
             self._gopkg_data[purl]["dependencies"].append(icm_dependency)
-        elif dependency.type == "gomod":
-            icm_source = {"purl": dependency.to_purl()}
-            self._gopkg_data[purl]["sources"].append(icm_source)
+
+    def set_go_package_sources(self):
+        """
+        Adjust source level dependencies for go packages.
+
+        Go packages are not related to Go modules in cachito DB. However, Go
+        sources are retreived in a per module basis. To set the proper source
+        in each content manifest entry, we associate each Go package to a Go
+        module based on their names.
+        """
+        for purl, pkg_data in self._gopkg_data.items():
+            if pkg_data["name"] in self._gomod_data:
+                self._gopkg_data[purl]["sources"] = self._gomod_data[pkg_data["name"]]
+            else:
+                previous_length = 0
+                for mod_name, sources in self._gomod_data.items():
+                    if pkg_data["name"].startswith(mod_name) and len(mod_name) > previous_length:
```

Comment:
To clarify, is the logic `len(mod_name) > previous_length` just to handle submodules/nested modules? If so, could you please add a comment explaining this for loop/if statement?